src/hashes/ternary/curl_p/regular/u256.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::{Index, IndexMut};
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub(super) struct U256([u64; 4]);
+
+impl Index<usize> for U256 {
+    type Output = u64;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        self.0.index(index)
+    }
+}
+
+impl IndexMut<usize> for U256 {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        self.0.index_mut(index)
+    }
+}
+
+impl U256 {
+    // Doing a bitwise AND with `1` produces a value between `0` and `1` which fits in an `i8` without truncation.
+    #[allow(clippy::cast_possible_truncation)]
+    pub(super) fn bit(&self, i: usize) -> i8 {
+        (self[(i / 64) % 4] >> (i % 64) & 1) as i8
+    }
+
+    pub(super) fn set_bit(&mut self, i: usize) {
+        self[(i / 64) % 4] |= 1 << (i % 64)
+    }
+
+    pub(super) fn shr_into(&mut self, x: &Self, shift: usize) -> &mut Self {
+        let offset = shift / 64;
+        let r = shift % 64;
+
+        if r == 0 {
+            for i in offset..4 {
+                self[(i - offset) % 4] |= x[i];
```

Comment:
That is how it was written in the go implementation. I agree that it is unnecessary in terms of code generation but at the same time I think it doesn't hurt having the immediate feedback of knowing that the index is not out of bounds by only reading this line instead of having to look up again what is `offset` and what are the possible `i` values and what-not.

Reply:
In Go this makes a difference as the compiler is not smart enough to detect that `(i-offset)` will never be out of bounds. After adding the `% 4` it becomes obvious to the compiler and saves us from rather costly runtime bound checks. I would guess that this would not be needed for the rust compiler...
However, if you think it improves readability then it is perfectly fine with me :+1: